src/coreclr/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -769,46 +851,23 @@ static void GenWalkFunc(void * context,
     } CONTRACT_END;
 
     GenerationTable *generationTable = (GenerationTable *)context;
+    generationTable->AddRecordNoLock(generation, rangeStart, rangeEnd, rangeEndReserved);
+    RETURN;
+}
 
-    _ASSERTE(generationTable->magic == GENERATION_TABLE_MAGIC);
-
-    ULONG count = generationTable->count;
-    if (count >= generationTable->capacity)
-    {
-        ULONG newCapacity = generationTable->capacity == 0 ? GenerationTable::defaultCapacity : generationTable->capacity * 2;
-        GenerationDesc *newGenDescTable = new (nothrow) GenerationDesc[newCapacity];
-        if (newGenDescTable == NULL)
-        {
-            // if we can't allocate a bigger table, we'll have to ignore this call
-            RETURN;
-        }
-        memcpy(newGenDescTable, generationTable->genDescTable, sizeof(generationTable->genDescTable[0]) * generationTable->count);
-        delete[] generationTable->genDescTable;
-        generationTable->genDescTable = newGenDescTable;
-        generationTable->capacity = newCapacity;
-    }
-    _ASSERTE(count < generationTable->capacity);
-
-    GenerationDesc *genDescTable = generationTable->genDescTable;
-
-    genDescTable[count].generation = generation;
-    genDescTable[count].rangeStart = rangeStart;
-    genDescTable[count].rangeEnd = rangeEnd;
-    genDescTable[count].rangeEndReserved = rangeEndReserved;
-
-    generationTable->count = count + 1;
+void GenerationTable::Refresh()
+{
+    // fill in the values by calling back into the gc, which will report
+    // the ranges by calling GenWalkFunc for each one
+    CrstHolder holder(&mutex);    
+    IGCHeap *hp = GCHeapUtilities::GetGCHeap();
+    hp->DiagDescrGenerations(GenWalkFunc, this);
```

Comment:
I'm worried about a sequence like
* Thread A: GenerationTable::Refresh takes the lock
* Thread B: An allocation that triggers a new region, the GC tries to call in to the profiler and add the region
* B tries to take lock and can't because A is holding it
* Now A and B are deadlocked because A is holding the lock waiting for B to finish allocating before doing the heap walk

Reply:
In my change, `GenerationTable::Refresh()` can be called in three places, during startup, before GC, and after GC.
The before and after GC callbacks are done within EE suspension.
So it is impossible to have a concurrent allocation happening while Refresh is happening.
The lock is only meant to protect against concurrent read (if that ever happens)